File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.js
function bfs(start) {
    const dist = new Array(N + 1).fill(-1);
    dist[start] = 0;
    const queue = [start];
    let front = 0;
    
    while (front < queue.length) {
        const v = queue[front++];
        for (const next of graph[v]) {
            if (dist[next] === -1) {
                dist[next] = dist[v] + 1;
                queue.push(next);
            }
        }
    }
    return dist.slice(1).reduce((acc, cur) => acc + cur, 0);
}

const input = require('fs').readFileSync(0).toString().trim().split('\n');
const [N, M] = input.shift().split(' ').map(Number);
const graph = Array.from({length: N + 1}, () => []);
for (let i = 0; i < M; i++) {
    const [a, b] = input.shift().split(' ').map(Number);
    graph[a].push(b);
    graph[b].push(a);
}

let minBacon = Infinity;
let result = 0;

for (let i = 1; i <= N; i++) {
    let bacon = bfs(i);
    
    if (bacon < minBacon) {
        minBacon = bacon;
        result = i;
    }
}

console.log(result);

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
from collections import deque
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

def bfs(start):
    visited = [-1] * (N + 1)
    visited[start] = 0
    queue = deque([start])
    while queue:
        v = queue.popleft()
        for next_v in graph[v]:
            if visited[next_v] == -1:
                visited[next_v] = visited[v] + 1
                queue.append(next_v)
    return sum(visited[1:])

min_bacon = float('inf')
result = 0

for i in range(1, N + 1):
    bacon = bfs(i)
    if bacon < min_bacon:
        min_bacon = bacon
        result = i

print(result)